source/fundamentals/data-formats/codecs.txt

```diff
@@ -11,38 +11,337 @@ Codecs
+- :java-docs:`DecoderContext <apidocs/bson/org/bson/codecs/DecoderContext.html>`
+- :java-docs:`BsonBinaryReader <apidocs/bson/org/bson/BsonBinaryReader.html>`
+
+.. _codecs-codecregistry:
 
 CodecRegistry
 -------------
 
-- Description
-- Example of setting Codec list
-- Example of getting a Codec
+A ``CodecRegistry`` is an immutable collection of ``Codec`` instances that
+encode and decode the Java classes they specify. You can use any of the
+following ``CodecRegistries`` class static factory methods to construct a
+``CodecRegistry`` from the ``Codec`` instances contained in the associated
+types:
+
+- ``fromCodecs()``
+- ``fromProviders()``
+- ``fromRegistries()``
+
+The following code snippet shows how to construct a ``CodecRegistry`` using
+the ``fromCodecs()`` method:
+
+.. code-block:: java
+   :copyable: true
+
+   CodecRegistry codecRegistry = CodecRegistries.fromCodecs(new IntegerCodec(), new PowerStatusCodec());
+
+In the example above, we assign the ``CodecRegistry`` the following ``Codec``
+implementations:
+
+- :java-docs:`IntegerCodec </apidocs/bson/org/bson/codecs/IntegerCodec.html>`,
+  a ``Codec`` that converts Integers and is part of the BSON package.
+- :ref:`PowerStatusCodec <codecs-powerstatus-codec>`, our sample ``Codec``
+  that converts certain Java strings to BSON booleans.
+
+You can retrieve the Codecs from the ``CodecRegistry`` instance from the
+prior example using the following code:
+
+.. code-block:: java
+   :copyable: true
+
+   Codec<String> powerStatus = codecRegistry.get(String.class);
+   Codec<Integer> integerCodec = codecRegistry.get(Integer.class);
+
+If you attempt to retrieve a ``Codec`` instance for a class that is not
+registered, the ``get()`` method throws a ``CodecConfigurationException``
+exception.
+
+For more information about the classes and interfaces in this section, see the
+following API documentation:
+
+- :java-docs:`CodecRegistries </apidocs/bson/org/bson/codecs/configuration/CodecRegistries.html>`
+- :java-docs:`IntegerCodec </apidocs/bson/org/bson/codecs/IntegerCodec.html>`
+
+.. _codecs-codecprovider:
 
 CodecProvider
 -------------
 
-- Description
-- Example of get method implementation
+A ``CodecProvider`` is an interface that contains abstract methods that create
+``Codec`` instances and assign them to a ``CodecRegistry`` instance. Similar
+to the ``CodecRegistry``, the BSON library uses the Codecs retrieved by
+the ``get()`` method to convert between Java and BSON data types.
+
+However, in cases in which you add a class that contains fields that require
+corresponding ``Codec`` instances, you need to ensure you construct the Codecs
+for the class' fields before the ``Codec`` for the class. You can use the
+``CodecProvider`` to add the Codecs to the registry before the one for which
+they are required.
+
+The following code example shows how you can implement ``CodecProvider`` to
+instantiate the Codecs on which the ``Monolight`` depends including the
```

Comment:
Should `Monolight` and `PowerStatusCodec` link to the section of the document where they are specified.

Reply:
That was my first thought, but when considering the following, I decided it was better not to do so for the following reasons:
1) Since we can't give links monospace formatting, we would need to provide additional information about the link as we do (or should do) with other links.
2) Keep the reader focused on what we're demonstrating which in this case doesn't offer information to understand the example other than maybe the type that getEncoderClass() returns in PowerStatusCodec. The content referenced in the link on line 201 features a link back to PowerStatusCodec including a description.

Let me know if that makes sense or whether you have a better solution, and whether this is a blocker for you.

Reply:
Thanks for the response. I think those reasons make sense. Going forward, I will try to critically assess linking to examples introduced elsewhere in the document as opposed to linking to referenced examples as a rule.